modules/basics/src/main/java/com/opengamma/basics/index/ImmutablePriceIndex.java

```diff
@@ -0,0 +1,514 @@
+/**
+ * Copyright (C) 2015 - present by OpenGamma Inc. and the OpenGamma group of companies
+ * 
+ * Please see distribution for license.
+ */
+package com.opengamma.basics.index;
+
+import java.io.Serializable;
+import java.util.Map;
+import java.util.NoSuchElementException;
+import java.util.Set;
+
+import org.joda.beans.Bean;
+import org.joda.beans.BeanDefinition;
+import org.joda.beans.ImmutableBean;
+import org.joda.beans.JodaBeanUtils;
+import org.joda.beans.MetaProperty;
+import org.joda.beans.Property;
+import org.joda.beans.PropertyDefinition;
+import org.joda.beans.impl.direct.DirectFieldsBeanBuilder;
+import org.joda.beans.impl.direct.DirectMetaBean;
+import org.joda.beans.impl.direct.DirectMetaProperty;
+import org.joda.beans.impl.direct.DirectMetaPropertyMap;
+
+import com.opengamma.basics.currency.Currency;
+import com.opengamma.basics.location.Country;
+import com.opengamma.basics.schedule.Frequency;
+
+/**
+ * A price index implementation based on an immutable set of rules.
+ * <p>
+ * A standard immutable implementation of {@link PriceIndex}.
+ * <p>
+ * In most cases, applications should refer to indices by name, using {@link PriceIndex#of(String)}.
+ * The named index will typically be resolved to an instance of this class.
+ * As such, it is recommended to use the {@code PriceIndex} interface in application
+ * code rather than directly referring to this class.
+ */
+@BeanDefinition(cacheHashCode = true)
```

Comment:
Beans that are hashmap keys should have cached hash codes. Over time, we'll see if that is most beans. Right now, we know this one will be a hash key.